ocaml/xapi/xapi_observer.ml

```diff
@@ -96,24 +210,67 @@ let create ~__context ~name_label ~name_description ~hosts ~attributes
     ~enabled ~attributes ~endpoints ~components ;
   ref
 
-let unregister ~__context ~self ~host:_ =
+let unregister_components ~__context ~self =
   let uuid = Db.Observer.get_uuid ~__context ~self in
-  Tracing.destroy ~uuid
+  List.iter (fun c ->
+      let module Forwarder = (val get_forwarder c : ObserverInterface) in
+      Forwarder.destroy ~__context ~uuid
+  )
+
+let unregister ~__context ~self ~host:_ =
+  Db.Observer.get_components ~__context ~self
+  |> observed_components_of
+  |> unregister_components ~__context ~self
 
 let destroy ~__context ~self = Db.Observer.destroy ~__context ~self
 
-let set_trace_log_dir dir =
-  Tracing.Export.Destination.File.set_trace_log_dir dir
+let set_trace_log_dir ~__context dir =
+  List.iter
```

Comment:
This pattern is repeated several times. Could we create the list of modules once? I am also a bit dubious about using first-class modules in so many places.

Reply:
I don't see a way to parameterise this more effectively.